### PR TITLE
Fixes #299

### LINKED ISSUE #299: Reduce restrictions of pure annotation
Right now, pure annotations always have to be written on the same line as the "func" keyword. We should permit the "pure" annotation to be at an arbitrary clause position (or at the beginning of the line of the function declaration).

### PR DESCRIPTION
**Pull request by Felalolf**
The "pure" annotation can now be at several positions in the specification.

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 1]
```diff
@@ -421,11 +421,30 @@ object Parser {
...
+      case class EnsuresClause(exp: PExpression) extends FunctionSpecClause
+      case object PureClause extends FunctionSpecClause
+
+      lazy val functSpecClause: Parser[FunctionSpecClause] = {
+        "requires" ~> expression <~ eos ^^ RequiresClause |
+        "ensures" ~> expression <~ eos ^^ EnsuresClause |
+        "pure" <~ eos ^^^ PureClause
       }
 
+      functSpecClause.* ~ "pure".? ^^ {
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Parser.scala:439]: is it necessary to have here `"pure".?`? Shouldn't this already be covered by another `functSpecClause`?

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 2]
```diff
@@ -421,11 +421,30 @@ object Parser {
...
+        case clauses ~ pure =>
+          val pres = clauses.collect{ case x: RequiresClause => x.exp }
+          val posts = clauses.collect{ case x: EnsuresClause => x.exp }
+          val isPure = pure.nonEmpty || clauses.contains(PureClause)
+          PFunctionSpec(pres, posts, isPure)
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Parser.scala:444]: shouldn't you attach positional information to `PFunctionSpec` here?

**Review by ArquintL — APPROVED** [review]
Two comments, otherwise looks good

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 2 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:444]: that is done by the parser framework

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 1 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:439]: So, yes it is necessary.

**Review by jcp19 — APPROVED** [review]
Looks fine to me. The only awkward thing to me is that it allows functions with multiple pure clauses, but we can argue that the `pure` modifier is idempotent.

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 1 above)
Comment (ArquintL) [reply src/main/scala/viper/gobra/frontend/Parser.scala:439]: I see, then maybe this might be a nicer solution: `repsep(functSpecClause, eos) <~ eos.?`

**Comment by ArquintL** [comment]
@jcp19 maybe you can negative the effect of the pure keyword when having two :P

**Pull request merged**